src/hotspot/share/nmt/vmatree.cpp

```diff
@@ -34,7 +34,7 @@ const char* VMATree::statetype_strings[3] = {
 };
 
 VMATree::SummaryDiff VMATree::register_mapping(position A, position B, StateType state,
-                                               const RegionData& metadata) {
+                                               const RegionData& metadata, bool copy_flag) {
```

Comment:
We need to think about what the behavior should be if the caller passes `copy_flag=true` but vmatree disagrees and thinks the caller should specify a flag. At the moment we then quietly take over the flag from the `metadata` argument, but I think we should assert, then. Something like "reserving without pre-existing range requires a valid NMT flag".

Tied into that is the question of what the flag in the `metadata` argument should be for `copy_flag=true`. I would argue that in that case, the flag should be `mtNone`. In other words, if I am the caller and I commit memory, I don't want to think about flags. VMATree should figure this out.

In fact, we could argue that `mtNone` means `copy_flag==true` and then we don't need the new `copy_flag` argument. In that case, reserving with `mtNone` means "no clue about the flag, you figure it out" and then it requires a pre-existing mapping. Are there valid use cases for reserving ranges with mtNone as a valid NMT flag?

Reply:
Asserting may make sense on debug builds, as you might be making a mistake, but we should have pre-defined behavior for production builds anyhow, no `ShouldNotReachHere()`. I think that valid behavior is to set the flag to whichever was provided in the call.


Another option is to have a more elaborate heuristic. If you uncommit `[a, b)` then find the first range which starts inside of `[a, b)` and take its memory flag, if none is found then give up and take `mtNone`.

The nicest behavior is an `O(n)` where we go through all regions in `[a, b)` and convert each to its reserved version with the flag, then uncommit with holes in the range would leave those holes unreserved. 

>Are there valid use cases for reserving ranges with mtNone as a valid NMT flag?

Only in the context of then later setting the memory flag, which I think is a pattern which should be avoided if possible.

Reply:
I'm only considering the usage of this flag as something which should be used by an `uncommit_region` API, if that wasn't clear.

Reply:
> Asserting may make sense on debug builds, as you might be making a mistake, but we should have pre-defined behavior for production builds anyhow, no ShouldNotReachHere(). I think that valid behavior is to set the flag to whichever was provided in the call.

I disagree :)

The point of this default mechanism is to take cognitive load from the programmer. I don't want to have to think about the flag when I rely on this mechanism. Having this mechanism, but then still forcing the caller to specify a flag "just in case you commit into unreserved territory" is not a big help.

By using copy_flag=true, I declare that I expect there to be a prior mapping in that range. That is part of the API contract. If that expectation is broken, I did something wrong. This is not something that can happen due to random runtime conditions. Therefore, like any other invalid state it should be handled with an assert. I don't think it makes a lot of sense to think about what to do in release builds: this is an error like many others, and we have many asserts in the JVM and we only rarely encode a valid non-assert path.

> Another option is to have a more elaborate heuristic. If you uncommit [a, b) then find the first range which starts inside of [a, b) and take its memory flag, if none is found then give up and take mtNone.

This does not sound very appealing tbh. In most cases, this would be an error. It is not NMTs responsibility to catch this error, but I would not go out of my way to accomodate it case either. 

> The nicest behavior is an O(n) algorithm where we go through all regions in [a, b) and convert each to its reserved version with the flag, then uncommit with holes in the range would leave those holes unreserved.

You lost me here. This is what we do right now, isn't it? VMATree just replaces all existing mappings between [a, b) with the new mapping?

> I'm only considering the usage of this flag as something which should be used by an uncommit_region API, if that wasn't clear.

Restricting this to uncommit would not use its full potential.

Consider this example: In the Metaspace allocator I would like to commit memory, but that memory can be tagged with either mtClassSpace or mtMetaspace, and in the course of Lilliput we may see even more tags. In all cases I reuse the same allocator code. Here, why should I be forced to have to specify the tag on commit? I already marked the underlying memory with the correct tag when reserving it.

The ability of painting regions with a certain flag and then having subsequent memory operations on that memory preserve the flag is powerful. You can, for example, in Metaspace account memory at the chunk level. Without having to add code everywhere that keeps track of those flags. You only have to paint the chunks you give to a Metaspace arena with the correct flag, and from then on the whole process is automatic.

Reply:
> I disagree :)
> 
> The point of this default mechanism is to take cognitive load from the programmer. I don't want to have to think about the flag when I rely on this mechanism. Having this mechanism, but then still forcing the caller to specify a flag "just in case you commit into unreserved territory" is not a big help.
> 
> By using copy_flag=true, I declare that I expect there to be a prior mapping in that range. That is part of the API contract. If that expectation is broken, I did something wrong. This is not something that can happen due to random runtime conditions. Therefore, like any other invalid state it should be handled with an assert. I don't think it makes a lot of sense to think about what to do in release builds: this is an error like many others, and we have many asserts in the JVM and we only rarely encode a valid non-assert path.

I don't understand what you disagree with specifically. I'm saying that asserting is fine, but we shouldn't give up and kill the VM in release builds with a `ShouldNotReachHere()` in case the programmer messed up when reporting diagnostics to NMT. It's not necessarily a bug in the actual 'business logic' of the code and we can recover from it by providing worse diagnostics.

> 
> 
>     The nicest behavior is an O(n) algorithm where we go through all regions in [a, b) and convert each to its reserved version with the flag, then uncommit with holes in the range would leave those holes unreserved.
> 
> You lost me here. This is what we do right now, isn't it? VMATree just replaces all existing mappings between [a, b) with the new mapping?

I was thinking in the sense of `uncommit`. Example:

```
reserve(0, 100, mtTest);
reserve(600, 1000, mtTest);
//  [0, 100) and [600, 1000) reserved with flag mtTest
commit(0, 100); commit(600, 1000);
//  [0, 100) and [600, 1000) committed with flag mtTest
// Now we can do
uncommit(0, 1000);
// And we end up with [0, 100) and [600, 1000) being reserved with flag mtTest
```

That is not equivalent to `register_mapping` right now, but `register_mapping` already does everything to support this kind of operation.




> 
>     I'm only considering the usage of this flag as something which should be used by an uncommit_region API, if that wasn't clear.
> 
> Restricting this to uncommit would not use its full potential.
> 
> Consider this example: In the Metaspace allocator I would like to commit memory, but that memory can be tagged with either mtClassSpace or mtMetaspace, and in the course of Lilliput we may see even more tags. In all cases I reuse the same allocator code. Here, why should I be forced to have to specify the tag on commit? I already marked the underlying memory with the correct tag when reserving it.
> 
> The ability of painting regions with a certain flag and then having subsequent memory operations on that memory preserve the flag is powerful. You can, for example, in Metaspace account memory at the chunk level. Without having to add code everywhere that keeps track of those flags. You only have to paint the chunks you give to a Metaspace arena with the correct flag, and from then on the whole process is automatic.

I don't disagree with this, I wasn't thinking about committing when I wrote my comments.

Reply:
> > I disagree :)
> > The point of this default mechanism is to take cognitive load from the programmer. I don't want to have to think about the flag when I rely on this mechanism. Having this mechanism, but then still forcing the caller to specify a flag "just in case you commit into unreserved territory" is not a big help.
> > By using copy_flag=true, I declare that I expect there to be a prior mapping in that range. That is part of the API contract. If that expectation is broken, I did something wrong. This is not something that can happen due to random runtime conditions. Therefore, like any other invalid state it should be handled with an assert. I don't think it makes a lot of sense to think about what to do in release builds: this is an error like many others, and we have many asserts in the JVM and we only rarely encode a valid non-assert path.
> 
> I don't understand what you disagree with specifically. I'm saying that asserting is fine, but we shouldn't give up and kill the VM in release builds with a `ShouldNotReachHere()` in case the programmer messed up when reporting diagnostics to NMT. It's not necessarily a bug in the actual 'business logic' of the code and we can recover from it by providing worse diagnostics.
> 
> > ```
> > The nicest behavior is an O(n) algorithm where we go through all regions in [a, b) and convert each to its reserved version with the flag, then uncommit with holes in the range would leave those holes unreserved.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > You lost me here. This is what we do right now, isn't it? VMATree just replaces all existing mappings between [a, b) with the new mapping?
> 
> I was thinking in the sense of `uncommit`. Example:
> 
> ```
> reserve(0, 100, mtTest);
> reserve(600, 1000, mtTest);
> //  [0, 100) and [600, 1000) reserved with flag mtTest
> commit(0, 100); commit(600, 1000);
> //  [0, 100) and [600, 1000) committed with flag mtTest
> // Now we can do
> uncommit(0, 1000);
> // And we end up with [0, 100) and [600, 1000) being reserved with flag mtTest
> ```

In VMAtree sure, but in reality not: since NMT has no complete knowledge about the address space. Uncommitting a range containing what NMT sees as address holes will wipe any mappings that happen to be in these holes that NMT does not know. Which are quite a lot.

> 
> That is not equivalent to `register_mapping` right now, but `register_mapping` already does everything to support this kind of operation.
> 
> > ```
> > I'm only considering the usage of this flag as something which should be used by an uncommit_region API, if that wasn't clear.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Restricting this to uncommit would not use its full potential.
> > Consider this example: In the Metaspace allocator I would like to commit memory, but that memory can be tagged with either mtClassSpace or mtMetaspace, and in the course of Lilliput we may see even more tags. In all cases I reuse the same allocator code. Here, why should I be forced to have to specify the tag on commit? I already marked the underlying memory with the correct tag when reserving it.
> > The ability of painting regions with a certain flag and then having subsequent memory operations on that memory preserve the flag is powerful. You can, for example, in Metaspace account memory at the chunk level. Without having to add code everywhere that keeps track of those flags. You only have to paint the chunks you give to a Metaspace arena with the correct flag, and from then on the whole process is automatic.
> 
> I don't disagree with this, I wasn't thinking about committing when I wrote my comments.

Reply:
The argument name `copy_flag` suggests to me that we will be using the provided `flag` from `metadata`, but the implementation does the reverse.

I think we might need a better name, ex: `use_implicit_flag` in which case we copy the flag out of the region itself, or we could use `use_explicit_flag` and do the opposite.

Reply:
I see what you're saying, maybe `use_found_flag` or `use_flag_inplace`? At least the assertions give a strong hint that that assumption is wrong. Better names are (almost) always welcome.